cpp/include/cudf/io/text/data_chunk_source_factories.hpp

```diff
@@ -0,0 +1,229 @@
+  };
+
+ public:
+  istream_data_chunk_reader(std::unique_ptr<std::istream> datastream)
+    : _datastream(std::move(datastream)), _buffers(), _tickets(2)
+  {
+    // create an event to track the completion of the last device-to-host copy.
+    for (uint32_t i = 0; i < _tickets.size(); i++) {
+      CUDA_TRY(cudaEventCreate(&(_tickets[i].event)));
+    }
+  }
+
+  ~istream_data_chunk_reader()
+  {
+    for (uint32_t i = 0; i < _tickets.size(); i++) {
+      CUDA_TRY(cudaEventDestroy(_tickets[i].event));
+    }
+  }
+
+  device_span<char> find_or_create_data(uint32_t size, rmm::cuda_stream_view stream)
+  {
+    auto search = _buffers.find(stream.value());
+
+    if (search == _buffers.end() || search->second.size() < size) {
+      _buffers[stream.value()] = rmm::device_buffer(size, stream);
+    }
+
+    return device_span<char>(static_cast<char*>(_buffers[stream.value()].data()), size);
+  }
+
+  device_span<char const> get_next_chunk(uint32_t read_size, rmm::cuda_stream_view stream) override
+  {
+    CUDF_FUNC_RANGE();
+
+    auto& h_ticket = _tickets[_next_ticket_idx];
+
+    _next_ticket_idx = (_next_ticket_idx + 1) % _tickets.size();
+
+    // synchronize on the last host-to-device copy, so we don't clobber the host buffer.
+    CUDA_TRY(cudaEventSynchronize(h_ticket.event));
+
+    // resize the host buffer as necessary to contain the requested number of bytes
+    if (h_ticket.buffer.size() < read_size) { h_ticket.buffer.resize(read_size); }
+
+    // read data from the host istream in to the pinned host memory buffer
+    _datastream->read(h_ticket.buffer.data(), read_size);
+
+    // adjust the read size to reflect how many bytes were actually read from the data stream
+    read_size = _datastream->gcount();
+
+    // get a view over some device memory we can use to buffer the read data on to device.
+    auto chunk_span = find_or_create_data(read_size, stream);
+
+    // copy the host-pinned data on to device
+    CUDA_TRY(cudaMemcpyAsync(  //
+      chunk_span.data(),
+      h_ticket.buffer.data(),
+      read_size,
+      cudaMemcpyHostToDevice,
+      stream.value()));
+
+    // record the host-to-device copy.
+    CUDA_TRY(cudaEventRecord(h_ticket.event, stream.value()));
+
+    // return the view over device memory so it can be processed.
+    return chunk_span;
+  }
+
+ private:
+  uint32_t _next_ticket_idx = 0;
+  std::unique_ptr<std::istream> _datastream;
+  std::unordered_map<cudaStream_t, rmm::device_buffer> _buffers;
+  std::vector<host_ticket> _tickets;
+};
+
+/**
+ * @brief a reader which produces view of device memory which represent a subset of the input device
+ * span
+ *
+ */
+class device_span_data_chunk_reader : public data_chunk_reader {
+ public:
+  device_span_data_chunk_reader(device_span<char const> data) : _data(data) {}
+
+  device_span<char const> get_next_chunk(uint32_t read_size, rmm::cuda_stream_view stream) override
```

Comment:
Why are we using `uint32_t` everywhere? `std::size_t` would be more canonical. 
```suggestion
  device_span<char const> get_next_chunk(std::size_t read_size, rmm::cuda_stream_view stream) override
```

Reply:
oversight. I'll get it cleaned up.